SwiftCompilerSources/Sources/Optimizer/Utilities/AccessUtils.swift

```diff
@@ -0,0 +1,421 @@
+      return (this.isExclusiveIndirectParameter || that.isExclusiveIndirectParameter) && this != that
+    case let (this as RefElementAddrInst, that as RefElementAddrInst):
+      return (this.fieldIndex != that.fieldIndex)
+    default:
+      if isUniquelyIdentified && other.isUniquelyIdentified && baseKind != other.baseKind { return true }
+      // property: `isUniquelyIdentified` XOR `isObject`
+      if isUniquelyIdentified && other.baseKind.isObject { return true }
+      if baseKind.isObject && other.isUniquelyIdentified { return true }
+      return false
+    }
+  }
+}
+
+/// An `AccessPath` is a pair of a `base: AccessBase` and a `projectionPath: Path`
+/// which denotes the offset of the access from the base in terms of projections.
+struct AccessPath : CustomStringConvertible {
+  var base: AccessBase
+
+  /// address projections only
+  var projectionPath: SmallProjectionPath
+
+  var description: String {
+    "\(projectionPath): \(base)"
+  }
+
+  func isDistinct(from other: AccessPath) -> Bool {
+    return base.isDistinct(from: other.base) ||
+      (base.baseAddress == other.base.baseAddress && projectionPath != other.projectionPath)
+  }
+}
+
+/// An `AccessStoragePath` is the reference (or a value for which the
+/// reference is a stored property) an address originates from.
+struct AccessStoragePath {
+  let storage: Value
+
+  /// Only valid paths are: `"<sequence of value projections>.<one reference projection>.<sequence of address projections>"`
+  let path: SmallProjectionPath
+}
+
+typealias AccessStorage = (maybeRoot: AccessStoragePath?, base: AccessBase)
+
+enum EnclosingScope {
+  case scope(BeginAccessInst)
+  case base(AccessBase)
+}
+
+
+func canBeOperandOfIndexAddr(_ value: Value) -> Bool {
+  switch value {
+  case is IndexAddrInst, is RefTailAddrInst:
+    return true
+  default:
+    return false
+  }
+}
+
+/// Given a `%addr = pointer_to_address %ptr_operand` instruction tries to identify
+/// the address the pointer operand `ptr_operand` originates from, if any exists.
+/// This is useful to identify patterns like
+/// ```
+/// %orig_addr = global_addr @...
+/// %ptr = address_to_pointer %orig_addr
+/// %addr = pointer_to_address %ptr
+/// ```
+/// which might arise when `[global_init]` functions for global addressors are inlined.
+///
+/// Alternatively, if the pointer originates from a ``FunctionArgument``, the argument is returned.
+///
+/// This underlying use-def traversal might cross phi arguments to identify the originating address
+/// to handle diamond-shaped control-flow with common originating address which might arise due to transformations ([example] (https://github.com/apple/swift/blob/8f9c5339542b17af9033f51ad7a0b95a043cad1b/test/SILOptimizer/access_storage_analysis_ossa.sil#L669-L705)) .
+struct PointerIdentification {
+  private var walker = PointerIdentificationUseDefWalker()
+
+  typealias AddressOrPointerArgument = Value
```

Comment:
Personally I don't like introducing this indirections/abstractions just for documentation purposes.
IMO, it produces more confusion than clarity.
Instead, just return a `Value?` and add a comment that this either returns an address value or a function argument.